interfaces/builtin/raw_volume.go

```diff
@@ -0,0 +1,103 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+// Only allow disk device partitions; not loop, ram, CDROM, generic SCSI,
+// network, tape, raid, etc devices
+const rawVolumeSummary = `allows access to disk block device partitions`
+
+const rawVolumeBaseDeclarationPlugs = `
+  raw-volume:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const rawVolumeBaseDeclarationSlots = `
+  raw-volume:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+// https://www.kernel.org/doc/Documentation/admin-guide/devices.txt
+// For now, only list common devices and skip the following:
+// /dev/mfm{a,b} rw,                        # Acorn MFM
+// /dev/ad[a-p] rw,                         # ACSI
+// /dev/pd[a-d] rw,                         # Parallel port IDE
+// /dev/pf[0-3] rw,                         # Parallel port ATAPI
+// /dev/ub[a-z] rw,                         # USB block device
+const rawVolumeConnectedPlugAppArmor = `
+# Description: Allow write access to disk block device partitions.
+
+@{PROC}/devices r,
+/run/udev/data/b[0-9]*:[0-9]* r,
+/sys/block/ r,
+/sys/devices/**/block/** r,
+
+# Access to raw devices, not individual partitions. Note that partition '0' is
+# not included since it refers to the whole disk.
+
+# IDE, MFM, RLL. 1-63 partitions
+/dev/hd[a-t]{[1-9],[1-5][0-9],6[0-3]} rw,
+
+# SCSI. 1-15 partitions
+/dev/sd{,[a-h]}[a-z]{[1-9],1[0-5]} rw,
+/dev/sdi[a-v]{[1-9],1[0-5]} rw,
+
+# I2O hard disk. 1-15 partitions
+/dev/i2o/hd{,[a-c]}[a-z]{[1-9],1[0-5]} rw,
+/dev/i2o/hdd[a-x]{[1-9],1[0-5]} rw,
+
+# MMC (up to 1000 devices). 1-7 partitions
+/dev/mmcblk[0-9]{,[0-9],[0-9][0-9]}p[1-7] rw,
```

Comment:
why we only allow 1 to 7 partitions for MMC?
often we have even 10+ partitions

Reply:
@kubiko - this comes from here:
```
  179 block       MMC block devices
		  0 = /dev/mmcblk0      First SD/MMC card
		  1 = /dev/mmcblk0p1    First partition on first MMC card
		  8 = /dev/mmcblk1      Second SD/MMC card
		    ...

		The start of next SD/MMC card can be configured with
		CONFIG_MMC_BLOCK_MINORS, or overridden at boot/modprobe
		time using the mmcblk.perdev_minors option. That would
		bump the offset between each card to be the configured
		value instead of the default 8.
```

I sounds like you bump the offset then. Should this be 16? 64?

Reply:
Quick check and Ubuntu kernel I could gather have CONFIG_MMC_BLOCK_MINORS set to 32
I seems like this rather dynamic value to restrict based on in
snapdragons devices are having regularly 10+ partitions
Same is actually also for /dev/sd*
On snapdragon 820 with UFS storage, one device has 33 partitions :)